packages/sdk/src/types/runtime/userGroupCollection.ts

```diff
@@ -0,0 +1,139 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { Context } from '.';
+
+// tslint:disable:no-bitwise
+
+/**
+ * A set of user group IDs. User groups are used to selectively enable several different
+ * properties of actors based on the memberships of the viewing user. All users not assigned
+ * a group are in the `default` group. See [[User.groups]], [[Appearance.enabled]].
+ */
+export default class UserGroupCollection extends Set<string> {
+    public static readonly ALL_PACKED = ~0;
+    public static readonly NONE_PACKED = 0;
+
+    private context: Context;
+
+    constructor(initialContents: Iterable<string> = null, context: Context = null) {
+        super(initialContents);
+        if (context) {
+            this.setContext(context);
+        }
+    }
+
+    public static All(context: Context) {
+        return this.FromPacked(context, this.ALL_PACKED);
+    }
+
+    public static FromPacked(context: Context, value: number): UserGroupCollection {
+        const group = new UserGroupCollection(null, context);
+        group.setPacked(value);
+        return group;
+    }
+
+    /** @hidden */
+    public setContext(c: Context) {
+        this.context = c;
+        for (const group of this) {
+            this.getOrAddMapping(group);
+        }
+    }
+
+    public packed() {
+        let pack = 0;
+        for (const group of this) {
+            pack |= this.getOrAddMapping(group);
+        }
+        return pack;
+    }
+
+    public toJSON() {
+        return this.packed();
+    }
+
+    private getOrAddMapping(name: string): number {
+        if (!this.context) {
+            throw new Error("Cannot serialize a user group without associating it with a context first");
+        }
+
+        const mapping = this.context.internal.userGroupMapping;
+        if (!mapping[name]) {
+            const lastIndex = Object.keys(mapping).length;
+            // all bitwise inputs are coerced to 32-bit signed ints
```

Comment:
If someone runs out of groups we can switch to a solution that uses a growable UInt32Array.

Reply:
My thoughts exactly. Or a byte array stored as a string or something. Lots of scaling options, but my intuition says that 31 groups should be plenty.

---
In reply to: [267573653](https://github.com/Microsoft/mixed-reality-extension-sdk/pull/270#discussion_r267573653) [](ancestors = 267573653)